### PR TITLE
Clean up data loading and randomization; add scheduler

### PR DESCRIPTION
**Pull request by pokey**
- Fixes https://github.com/chaosparrot/parrot.py/issues/18
- Fixes https://github.com/chaosparrot/parrot.py/issues/15

### CODE REVIEW

File: lib/audio_net.py [hunk 1]
```diff
@@ -94,23 +100,20 @@ def __init__(self, dataset, net_count = 1, audio_settings = None):
         
         split = int(np.floor(self.validation_split * self.dataset_size))
 
+        # Split the dataset into validation and training data loaders
+        dataset_rng = torch.Generator().manual_seed(self.data_seed)
+        train_dataset, val_dataset = torch.utils.data.random_split(dataset, [self.dataset_size - split, split], generator=dataset_rng)
```
Comment (pokey) [inline lib/audio_net.py:105]: Note that dataset now has its own seed, and we do the split once, rather than doing it once per ensemble member

File: lib/audio_net.py [hunk 2]
```diff
@@ -94,23 +100,20 @@ def __init__(self, dataset, net_count = 1, audio_settings = None):
...
-            
-            train_sampler = SubsetRandomSampler(self.train_indices[i])
-            valid_sampler = SubsetRandomSampler(val_indices)
-            self.train_loaders.append(torch.utils.data.DataLoader(dataset, batch_size=self.batch_size, sampler=train_sampler))
-            self.validation_loaders.append(torch.utils.data.DataLoader(dataset, batch_size=self.batch_size, sampler=valid_sampler))
+            optimizer = optim.SGD(self.nets[i].parameters(), lr=0.003, momentum=0.9, nesterov=True)
+            self.optimizers.append(optimizer)
+            self.schedulers.append( torch.optim.lr_scheduler.ReduceLROnPlateau(optimizer, 'min'))
+
+            self.train_loader = torch.utils.data.DataLoader(train_dataset, batch_size=self.batch_size, shuffle=True)
```
Comment (pokey) [inline lib/audio_net.py:115]: We create a shuffling data loader, so we don't need to do shuffling during training; that happens for us automatically

File: lib/audio_net.py [hunk 3]
```diff
@@ -94,23 +100,20 @@ def __init__(self, dataset, net_count = 1, audio_settings = None):
...
-            train_indices, val_indices = indices[split:], indices[:split]
-            self.train_indices.append( train_indices)
-            
-            train_sampler = SubsetRandomSampler(self.train_indices[i])
-            valid_sampler = SubsetRandomSampler(val_indices)
-            self.train_loaders.append(torch.utils.data.DataLoader(dataset, batch_size=self.batch_size, sampler=train_sampler))
-            self.validation_loaders.append(torch.utils.data.DataLoader(dataset, batch_size=self.batch_size, sampler=valid_sampler))
+            optimizer = optim.SGD(self.nets[i].parameters(), lr=0.003, momentum=0.9, nesterov=True)
+            self.optimizers.append(optimizer)
+            self.schedulers.append( torch.optim.lr_scheduler.ReduceLROnPlateau(optimizer, 'min'))
```
Comment (pokey) [inline lib/audio_net.py:113]: Added a scheduler

**Comment by chaosparrot** [comment]
I initially placed a different validation set / train set on each net to make sure each net would see different data ( thus performing better as an ensemble than if they all saw the same data ). Otherwise the only variation they would have after many epochs would be the starting position right?

My question would be: since 3 nets using different train / validation sets can use more data for their training, wouldn't they perform better on novel data? Given a testset which none of the models have seen, wouldn't 3 ensembled models with different training data perform better on than an ensemble created where each model saw exactly the same data?

( From a combined ensemble validation score I can definitely see the benefits of a single split train / validation set, because with an ensemble where the validation set has been seen by some of the models there's obvious data pollution and the results will be skewed )

**Comment by pokey** [comment]
Huh interesting idea. Makes me a bit nervous, eg we'd want to make sure we know which data seed each ensemble member was using in case we want to resume from checkpoint. But I guess could work?

@ym-han any thoughts? Is this something you've seen before? Reminds me of k-fold cross validation tbh, tho not exactly the same

**Comment by ym-han** [comment]
I haven't looked at the code so I can't be sure, but it sounds like this could be bagging (or something similar). There's some discussion and a link to some references on the sklearn bagging classifier page (https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.BaggingClassifier.html), and which I'm going to quote:

> A Bagging classifier is an ensemble meta-estimator that fits base classifiers each on random subsets of the original dataset and then aggregate their individual predictions (either by voting or by averaging) to form a final prediction. Such a meta-estimator can typically be used as a way to reduce the variance of a black-box estimator (e.g., a decision tree), by introducing randomization into its construction procedure and then making an ensemble out of it.

> This algorithm encompasses several works from the literature. When random subsets of the dataset are drawn as random subsets of the samples, then this algorithm is known as Pasting [[1]](https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.BaggingClassifier.html#rb1846455d0e5-1). If samples are drawn with replacement, then the method is known as Bagging [[2]](https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.BaggingClassifier.html#rb1846455d0e5-2). When random subsets of the dataset are drawn as random subsets of the features, then the method is known as Random Subspaces [[3]](https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.BaggingClassifier.html#rb1846455d0e5-3). Finally, when base estimators are built on subsets of both samples and features, then the method is known as Random Patches [[4]](https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.BaggingClassifier.html#rb1846455d0e5-4).

Breiman 1996 "Bagging Predictors" and the chapter on bagging in Richard Berk's _Statistical Learning from a Regression Perspective_ seem useful if you want to read up more.

**Comment by pokey** [comment]
Yeah that sounds right. But I presume all of those methods assume you're sampling the training data with a fixed held out validation set, whereas here the validation set of one ensemble member is used as training data for another

**Comment by chaosparrot** [comment]
Did some reading on KFold and cross validation and Pokey is right in the sense that there is a held out set kept separate from the validation sets for each model. I.e. given we have a data set of A, B, C ... Z, and two models, these are the current splits ( note the overlap ):

Model     | Training | Validation
-------------|--------------|---------------
Model A | A, B, C, D, E, F, G, H, I, O, P, Q, R, |  J K Y L M N Z
Model B | N, O, P, Q, R, S, T, U, V, F, G, H, I, | W X Y Z D E L
Ensemble | - | J K L M N O W X Y Z D E

Whereas the, in my opinion, best split would be:

Model     | Training | Validation
-------------|--------------|---------------
Model A | A, B, C, D, E, F, G, H, I, O, P, Q, R, |  J K L M N O
Model B | N, O, P, Q, R, S, T, U, V, F, G, H, I, | A B C D E F
Ensemble | - | V W X Y Z

Where we keep 10 percent of the total data set held out to test the ensemble on, use the remaining 90 percent for training and validation, and have the models have validation sets that do not overlap with one another.

The issue then is still that we need to keep the random seed persisted to keep the split available for checkpointing, but I think this would generate the best model results that do not have any data pollution going on in the validation sets.

**Comment by ym-han** [comment]
This table is helpful. Yes, I agree that it's important that whatever data is in the test set for the ensemble is _not_ data that has been used to either train or tune the hyperparams of any of the models of the ensemble.

**Comment by pokey** [comment]
hmm I'd be tempted to either
- go with a simple / standard ensemble setup like the one proposed in this PR, or
- stick with the existing approach (with some tweaks for reproducibility eg capturing seeds).  I can see the argument for utilising all available data for the ensemble, as the average user isn't publishing a paper or anything so doesn't really need proper held-out test set

Happy to hash this one out on Discord tho

**Comment by pokey** [comment]
Ok I'll close this one for now; at some point prob worth cleaning up the split code and pulling the LR schedule stuff out of here, but this PR would need a lot of tweaking to get there.  I pointed to this PR from the relevant issues
